src/Type/Php/SubstrDynamicReturnTypeExtension.php

```diff
@@ -53,7 +108,28 @@ public function getTypeFromFunctionCall(
 			}
 		}
 
-		return new StringType();
+		// handle fatal error cases
+		if ($string instanceof ArrayType || $string instanceof ObjectType) {
+			if ($this->phpVersion->getVersionId() >= 80000) {
+				return new NeverType();
+			}
+			return new NullType();
+		}
+
+		// error case
+		if (!$string instanceof StringType && !$string instanceof FloatType) {
+			return $errorType;
+		}
+
+		// happy path
+		if ($this->phpVersion->getVersionId() >= 80000) {
+		    return new StringType();
+	    }
+
+		return TypeCombinator::union(
+			new StringType(),
+			new ConstantBooleanType(false)
```

Comment:
IMO this is the only controversial case: returning `string|false` leads to ergonimc issues in substr() calling code.
Most codebases I know handle substr() like weakly typed - they don‘t handle a possible false-return.

we also see in the phpstan-src codebase its handled like that